rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -178,38 +177,61 @@ class DebruijnInterpreter[M[_], F[_]](
       ReduceError(
         s"The number of terms in the Par is ${terms.size}, which exceeds the limit of ${termSplitLimit}."
       ).raiseError[M, Unit]
-    else
-      terms.zipWithIndex.toList.parTraverse_ {
-        case (term, index) =>
-          eval(term)(env, split(index))
+    else {
+
+      // Collect errors from all parallel execution paths (pars)
+      terms.zipWithIndex.toVector
+        .parTraverse {
+          case (term, index) =>
+            eval(term)(env, split(index))
+              .map(_ => none[Throwable])
+              .handleError(_.some)
+        }
+        .map(_.flattenOption)
+        .flatMap(aggregateEvaluatorErrors)
+    }
+  }
+
+  private def aggregateEvaluatorErrors(errors: Vector[Throwable]) = errors match {
+    // No errors
+    case Vector() => ().pure[M]
+
+    // Out Of Phlogiston error is always single
+    // - if one execution path is out of phlo, the whole evaluation is also
+    case errList if errList.contains(OutOfPhlogistonsError) =>
+      OutOfPhlogistonsError.raiseError[M, Unit]
```

Comment:
These `raiseError` calls are the problem. The semantics of `parTraverse` when `raiseError` is called in an execution branch is not "wait for all threads running in parallel with this one to cancel before returning an error." It is, "run a cancellation process (`Task`) concurrently with the currently executing branches and return error immediately." In other words, `parTraverse` can return with an error before all threads are cancelled. Since it can return early, we can move past the point in the call sequence where we're able to handle errors.

The reason that we were getting an `OOPE` that would cause node to hang is because, given two branches `P | Q`, if `P` ran out of gas, it would throw an `OOPE`, but the result of `parTraverse` would return before `Q` was cancelled. As we exited the interpreter (and all opportunities for handling interpreter errors had passed), `Q` was still uncancelled. Then, `Q` would make _another_ call to `charge` and then throw _another_ `OOPE`. However, by the time the second `OOPE` was thrown, we'd already exited the interpreter, so instead of being caught and handled, it became an _uncaught_ exception, and was reported to the uncaught exception handler all the way at the top of node. We'd then try to reprocess the block, and the cycle continued.

Reply:
You'll be able to see this if you run the first test in `InterpreterSpec` with a large number of terms in parallel with many COMM events and a single error-throwing process on at least the second level (inside a continuation). I suggest you do that before you continue work on this PR, and before you even do that, please review the PR I opened so that it can be merged. If it turns out you're right, you can always replace what I've written with what you've written.

Reply:
@JosDenmark I completely agree with you. If failed _Tasks_ created by _parTraverse_ are not handled, many different scenarios can happen that are not desirable.
But _parTraverse_ is not the culprit, it's just the higher level abstraction as you already pointed out in the comment for _eval_ method where _parTraverse_ is used.

What I suggest is let's cut it at the source and not care about it anywhere else in our code base. Here is all that is needed for a fix. Result from each _Task_ spawned for each call to _eval_ cannot escape with unhanded error.
```scala
// Collect errors from all parallel execution paths (pars)
terms.zipWithIndex.toVector
  .parTraverse {
    case (term, index) =>
      eval(term)(env, split(index))
        .map(_ => none[Throwable])
        // This will prevent for Task to finish with an error
        // so failed Task cannot occur as a result of parTraverse
        .handleError(_.some)
  }
  .map(_.flattenOption)
  .flatMap(aggregateEvaluatorErrors)
```
There is no need to guard any other function separately if we know the problem can happen only when a new Task is spawned/exited. We can also implement this logic in the task scheduler which can do this for us so even using _parTraverse_ can be opaque to this inconvenience.
Examples of this pattern exists in JS [AggregateError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError) and C# [AggregateException](https://docs.microsoft.com/en-us/dotnet/api/system.aggregateexception).

In test you are proposing, with many terms and comm events, all parallel Tasks will use this code to handle all failed branches on any nesting level and execution will continue after the last _Task_ will produce result, `None` for success and `Some(Throwable)` for failed. So child Task cannot exit with the failure and all errors are aggregated and propagated to the parent.

In support of this, I've found a bug that is still undetected in your PR when error in user code does not consume all phlos in [RuntimeManagerTest](https://github.com/rchain/rchain/blob/890ffbb20/casper/src/test/scala/coop/rchain/casper/util/rholang/RuntimeManagerTest.scala#L605).